### DISCUSSION TITLE
Count numer of certain characters in input

### DISCUSSION THREAD

**Discussion by k0201** [taocpp/PEGTL#356] [Q&A] [answered]
Is there a possibility to  create a rule that passes if amount of certain characters exist in the input in the same number of times?
I am creating a parser that includes vector expressions such as [ x y z ] [ a b c ]. I did manage to parse if vectors are closed by ` if_must<one<'['>, until<one<']'>, allAcceptableNames>>`. However later in code I want to parse '[' and ']' separately.  This leads me to a problem that if I insert `opt<one<']'>>` in my grammar, an input containing only `']'` gets through. 
My idea was to count the number of `'['` and` ']'` and if is the same the input is correct. Is there a _simple_ way to do it?

**Comment by d-frey** [comment]
It probably depends on what you mean by "simple" and it is not fully clear to me, what you consider correct input. Is `][` a correct input? What about `[]][[]`? Both have an equal number of opening and closing symbols, but I think there are probably more rules to follow, like proper nesting. That said, a "simple" grammar should be sufficient, something like:

```c++
struct element : identifier {}; // or whatever else you need/want
struct part : seq< one<'['>, list< pad< element, space >, one<']'> >, opt<star<space>, part> > {};
```

You can then use the `at<>`-rule to do a look-ahead if needed, then add your "real" parsing run. Does this help?

**Reply by k0201** [reply]
Thank you for the reply.
So here is more info:
The parser is for XSPICE modifiers, there are only few allowed characters such as `~ [ ]`  and modifiers starting with `%` such as `%d %vd` . With them `?` stand as names of ports that can be 'modified'.
Only `[  ]` is a valid input, however there can be input without any `[ ]`, nested vectors and unclosed vectors are not allowed: `[ [ ] ]` and `][`  or `[]][[]` is not valid. 

> You can then use the at<>-rule to do a look-ahead if needed, then add your "real" parsing run.

I did find this idea and already using it, but a simple `? ] ` gets though. 

The code I have written is here:
https://gitlab.com/k0201pl/pegtl-eeschema-xspice-pattern-parser/-/blob/main/XSPICEpattern.h?ref_type=heads

And the problem I can't deal is I want have all the parsed names in `nodeSequence`  so I insert `[`  and `]` there and it gets parsed. 
Should I remove `[ ]` and make a separate selector just for them?

**Reply by d-frey** [reply]
I don't know anything about XSPICE, I'm afraid I can't help with any specifics.

From a quick glance it seems that you might need to rethink the grammar from the ground up and clearly formulate the actual grammar you want to parse, not just "this works partly, let's do the rest another time". For example, I'm very suspicious of `(` and `)` being simple separators that could occur anywhere and where those brackets never match and don't have any meaning besides being separators. There are other suspicious things like the `one<'~'>` not being used with `opt<...>`, but in a seemingly more complicated way.

**Reply by k0201** [reply]
Thanks, ok this kind-of answers my question. I see it will be easier to rewrite the grammar as there isn't a way to just write something like if( count('[') == count(']') ).

**Reply by k0201** [reply, marked as answer]
Ok,  I will close it, as the question is answered.
